katcp/server.py

```diff
@@ -2318,6 +2338,99 @@ def inform_callback(sensor, reading):
         yield gen.moment
         raise gen.Return(req.make_reply("ok", name, strategy, *params))
 
+    @gen.coroutine
+    def _handle_bulk_set_sensor_sampling(self, req, msg):
+        if not msg.arguments:
+            raise FailReply("No sensor name given.")
+
+        # Arguments start with one or more sensor names.  Check the first one now.
+        # A sensor object is required for testing the sampling strategies, so
+        # the first one is as good as any.
+        name = msg.arguments[0]
+        if name not in self._sensors:
+            raise FailReply("Unknown sensor name: %s." % name)
+        sensor = self._sensors[name]
+
+        # The client connection that is not specific to this request context
+        client = req.client_connection
+        katcp_version = self.PROTOCOL_INFO.major
+        format_inform = (format_inform_v5
+                         if katcp_version >= SEC_TS_KATCP_MAJOR
+                         else format_inform_v4)
+
+        def inform_callback(sensor, reading):
+            """Inform callback for sensor strategy."""
+            timestamp, status, value = reading
+            cb_msg = format_inform(sensor, timestamp, status, value)
+            client.inform(cb_msg)
+
+        # Parse arguments from the back, trying all possible strategies until
+        # a valid one is found.  The number of arguments to define a strategy
+        # is one for the name, plus the number of params expected for that type
+        # of strategy.
```

Comment:
Unfortunately a space is the katcp delimiter, but a comma might do if we update the KATCP spec for 5.1 to exclude comma from sensor-names, as in practice I don't think we have any sensors with comma's in the name.

The katcp spec has the following for the message BNF:
```
<message> ::= <type> <name> [<msgid>] <arguments> <eol>
<type> ::= "?" | "!" | "#"
<name> ::= alpha (alpha | digit | "-")*
<msgid> ::= "[" (digit / "0") digit* "]"
<whitespace> ::= (space | tab) [<whitespace>]
<eol> ::= newline | carriage-return
<arguments> ::= (<whitespace> <argument> <arguments>) | <whitespace> | ""
<argument> ::= (<plain> | <escape>)+
<escape> ::= "\" <escapecode>
<escapecode> ::= "\" | "_" | zero | "n" | "r" | "e" | "t" | "@"
<special> ::= backslash | space | null | newline | carriage-return | escape | tab
<plain> ::= character / <special>
```

Reply:
It just defines "character" (which I don't think is defined in standard BNF) without describing it further

Reply:
You can still include a space in an argument, it just has to be escaped. So on the wire it would look something like
```
?sensor-sampling sensor-1\_sensor-2\_sensor-3 period 500
```

Reply:
Yes true

Reply:
I agree that having to try all the strategies backwards isn't a great solution, and does prevent variable arguments in future.  Passing all the names as the first parameters with a separator character sounds like a much a better idea.  I prefer a comma to an escaped space.  It is much more readable and easier to type in debug session.  It seems almost as unlikely as a space in any existing sensor names. 
```
?sensor-sampling sensor-1,sensor-2,sensor-3 period 500
```
Are commas OK?
Question 2:  should we update the spec's section on sensors to include a list of allowed characters, and if so would that require an increment in the major version number?

Reply:
I think commas are probably okay. I think the spec generally needs more written about acceptable character sets, but I see that as a clarification rather than a major change - putting weird characters in sensor names is probably already going to break something somewhere.

Reply:
Thanks, then I will go with commas.

Reply:
Does anyone know where the LaTeX source files are for the KATCP guidelines doc?  Maybe @brickZA's memory goes back to 2012, when they were last updated?!

Reply:
It lives in https://katfs.kat.ac.za/svnSystemEngineering/General/KATCP/NRF-KAT7-6.0-IFCE-002%20(GuidelinesForCommunicationWithDevices)
I think "trunk" has the master source and "tags" the issued versions.

Reply:
I agree with @bmerry that we can handle the update as a clarification of valid characters in sensor names as a minor version change, not major.

Reply:
@bmerry and @lvdheever Proposed updates to the guidelines are here:  ska-sa/katcp-guidelines/pull/1
Please have a look.